grease/src/Grease/Shape.hs

```diff
@@ -314,3 +340,270 @@ deriving instance (ShowExt ext tag, ShowF tag) => Show (ArgShapes ext tag tys)
+-- for some of them, calculate a new list of arguments.
+replaceShapes ::
+  forall ext w tys.
+  ExtShape ext ~ PtrShape ext w =>
+  HasPtrWidth w =>
+  -- | Argument names
+  Ctx.Assignment (Const.Const String) tys ->
+  -- | Initial arguments
+  ArgShapes ext NoTag tys ->
+  -- | Replacement arguments
+  ParsedShapes ext ->
+  Either TypeMismatch (ArgShapes ext NoTag tys)
+replaceShapes names (ArgShapes args) (ParsedShapes replacements) =
+  -- TODO: Check that all the map keys are expected
+  ArgShapes
+    <$> Ctx.zipWithM (\(Const.Const nm) s -> replaceOne nm s) names args
+ where
+  replaceOne :: String -> Shape ext NoTag t -> Either TypeMismatch (Shape ext NoTag t)
+  replaceOne nm s =
+    case Map.lookup (Text.pack nm) replacements of
+      Just (C.Some replace) ->
+        let ty = shapeType ptrShapeType s
+         in let ty' = shapeType ptrShapeType replace
+             in case testEquality ty ty' of
+                  Just Refl -> Right replace
+                  Nothing ->
+                    Left $
+                      TypeMismatch
+                        { typeMismatchName = nm
+                        , expectedType = C.Some ty
+                        , foundType = C.Some ty'
+                        }
+      Nothing -> Right s
+
+isInSubProg ::
+  -- | the program counter from the base of the target ELF
+  Word64 ->
+  Subprogram ->
+  Bool
+isInSubProg w sub =
+  let entryMatch = (==) w <$> MDwarf.subEntryPC sub
+      def = fromMaybe False
+   in def
+        ( do
+            sdef <- MDwarf.subDef sub
+            lpc <- MDwarf.subLowPC sdef
+            hpc <- MDwarf.subHighPC sdef
+            pure $ w >= lpc && w < (lpc + hpc)
+        )
+        || def entryMatch
+
+rightToMaybe :: Either a b -> Maybe b
+rightToMaybe (Left _) = Nothing
+rightToMaybe (Right b) = Just b
+
+extractType :: Subprogram -> MDwarf.TypeRef -> Maybe MDwarf.TypeApp
+extractType sprog vrTy =
+  do
+    let mp = MDwarf.subTypeMap sprog
+    (mbtypeApp, _) <- Map.lookup vrTy mp
+    rightToMaybe mbtypeApp
+
+constructPtrTarget ::
+  HasPtrWidth w =>
+  TypeUnrollingBound ->
+  Subprogram ->
+  MDwarf.TypeApp ->
+  StateT VisitState Maybe (PtrTarget w NoTag)
+constructPtrTarget tyUnrollBound sprog tyApp =
+  ptrTarget Nothing <$> shapeSeq tyApp
+ where
+  ishape w = lift $ Just $ Seq.singleton $ Initialized NoTag (toBytes w)
+  padding :: (Integral a) => a -> MemShape w NoTag
+  padding w = Uninitialized (toBytes w)
+  -- if we dont know where to place members we have to fail/just place some bytes
+  buildMember :: HasPtrWidth w => Word64 -> MDwarf.Member -> StateT VisitState Maybe (Word64, Seq.Seq (MemShape w NoTag))
+  buildMember loc member =
+    do
+      memLoc <- lift $ MDwarf.memberLoc member
+      padd <- lift $ Just (if memLoc == loc then Seq.empty else Seq.singleton (padding $ memLoc - loc))
+      memberShape <- shapeOfTyApp $ MDwarf.memberType member
+      let memByteSize = sum $ memShapeSize ?ptrWidth <$> memberShape
```

Comment:
Since this is a sequence, perhaps
```suggestion
      memberShapes <- shapeOfTyApp $ MDwarf.memberType member
      let memByteSize = sum $ memShapeSize ?ptrWidth <$> memberShapes
```